presto-spi/src/main/java/com/facebook/presto/spi/ConsistentHashingNodeProvider.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+import static com.facebook.presto.spi.StandardErrorCode.NO_NODES_AVAILABLE;
+import static java.util.Collections.unmodifiableList;
+
+public class ConsistentHashingNodeProvider<T>
+        implements NodeProvider<T>
+{
+    private final SortedMap<Integer, T> circle = new TreeMap<>();
```

Comment:
I'm curious, if we create a new `TreeMap` for every query, how do we make sure hashing is consistent across queries?

Reply:
If I understand correctly, the node would always be put to the same place in the hash ring
(we can do optimization with nodeManager to only build the hash ring once, it's would be in the next step)